eth/watchers/roundswatcher.go

```diff
@@ -0,0 +1,160 @@
+}
+
+type Watcher interface {
+	Subscribe(sink chan<- []*blockwatch.Event) event.Subscription
+}
+
+// NewRoundsWatcher creates a new instance of RoundsWatcher and sets the initial cache through an RPC call to an ethereum node
+func NewRoundsWatcher(watcher Watcher, lpEth eth.LivepeerEthClient, addr ethcommon.Address) (*RoundsWatcher, error) {
+	dec, err := NewEventDecoder(addr, contracts.RoundsManagerABI)
+	if err != nil {
+		return nil, fmt.Errorf("error creating decoder: %v", err)
+	}
+
+	return &RoundsWatcher{
+		quit:    make(chan struct{}),
+		watcher: watcher,
+		lpEth:   lpEth,
+		dec:     dec,
+	}, nil
+}
+
+// LastInitializedRound gets the last initialized from cache
+func (rw *RoundsWatcher) LastInitializedRound() (*big.Int, error) {
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	return rw.lastInitializedRound, nil
+}
+
+// BlockHashForRound gets the blockhash for the last initialized round from cache
+func (rw *RoundsWatcher) BlockHashForRound(round *big.Int) ([32]byte, error) {
+	if round == nil {
+		return [32]byte{}, fmt.Errorf("round is nil")
+	}
+	lastRound, _ := rw.LastInitializedRound()
+	if lastRound.Cmp(round) != 0 {
+		return rw.lpEth.BlockHashForRound(round)
+	}
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	return rw.blockHashForRound, nil
+}
+
+// setLastInitializedRound sets the last initialized round in cache
+func (rw *RoundsWatcher) setLastInitializedRound(round *big.Int) {
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	rw.lastInitializedRound = round
+}
+
+//setBlochashForRound sets the blockhash for the last initialized round in cache
+func (rw *RoundsWatcher) setBlockHashForRound(hash [32]byte) {
+	rw.mu.RLock()
+	defer rw.mu.RUnlock()
+	rw.blockHashForRound = hash
+}
+
+// Watch the blockwatch subscription for NewRound events
+func (rw *RoundsWatcher) Watch() {
+	lr, err := rw.lpEth.LastInitializedRound()
+	if err != nil {
+		glog.Errorf("error fetching initial lastInitializedRound value: %v", err)
+	}
+	rw.setLastInitializedRound(lr)
+
+	bh, err := rw.lpEth.BlockHashForRound(lr)
+	if err != nil {
+		glog.Errorf("error fetching initial blockHashForRound value: %v", err)
+	}
+	rw.setBlockHashForRound(bh)
+
+	events := make(chan []*blockwatch.Event)
+	sub := rw.watcher.Subscribe(events)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case <-rw.quit:
+			return
+		case err := <-sub.Err():
+			glog.Error(err)
+		case events := <-events:
+			rw.handleBlockEvents(events)
+		}
+	}
+}
+
+// Stop watching for NewRound events
+func (rw *RoundsWatcher) Stop() {
+	close(rw.quit)
+}
+
+func (rw *RoundsWatcher) handleBlockEvents(events []*blockwatch.Event) {
+	for _, event := range events {
+		for _, log := range event.BlockHeader.Logs {
+			nr, err := rw.handleLog(log)
+			if err != nil {
+				glog.Error(err)
+			}
+			if event.Type == blockwatch.Added && nr.Round != nil {
+				rw.setLastInitializedRound(nr.Round)
+				rw.setBlockHashForRound(nr.BlockHash)
+			} else if event.Type == blockwatch.Removed && nr.Round != nil {
```

Comment:
Is `nr.Round` ever nil besides when there is an error unpacking a log?

Reply:
No, this is a leftover from when I was using abi.Unpack rather than the event decoder. Can be deleted.

Reply:
fixed in 80d3e25